File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <functional>

using namespace std;

void dfs(const vector<int>&numbers, int target, int& answer, int idx, int total){
    if(idx==numbers.size()){
        if(total==target){
            answer+=1;
        }
        return;
    }
    dfs(numbers,target,answer,idx+1,total+numbers[idx]);
    dfs(numbers,target,answer,idx+1,total-numbers[idx]);
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    dfs(numbers,target,answer,0,0);
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int answer = 0;
    public int solution(int[] numbers, int target) {
        
        dfs(0,numbers,target,0);
        
        return answer;
    }
    
    public void dfs(int result, int[] numbers, int target, int idx){
        if(idx==numbers.length){
            if(target==result){
                answer++;
            }
            return;
        }
        dfs(result+numbers[idx],numbers,target,idx+1);
        dfs(result-numbers[idx],numbers,target,idx+1);
    }
}